### ISSUE TITLE
How to query based on ObjectId?

### ISSUE THREAD

**Issue by mzdk100** [WithSecureLabs/mongo-rs#9]
When I use Filter to query data, how do I use the ObjectId within the database for querying?

**Comment by alexkornitzer** [comment]
It is currently not possible via derives, they would need to be expanded to support this. It can be done manually by implementing the traits though.

**Comment by mzdk100** [comment]
Can you give an example?

**Comment by alexkornitzer** [comment]
Sure, if its purely the filter that you want then it would be something like this:

```rust
use futures::stream::StreamExt;
use mongod::bson::{oid::ObjectId, Document};
use mongod::ext;
use mongod::{AsFilter, AsUpdate, Comparator, Error, Filter, Updates};
use mongod::{Bson, Mongo};

#[derive(Debug, Bson, Mongo)]
#[mongo(collection = "users", field, update)]
pub struct User {
    pub name: String,
    pub age: Option<u32>,
}

#[derive(Default)]
pub struct UserFilter {
    pub _id: Option<ObjectId>,
    pub name: Option<Comparator<String>>,
    pub age: Option<Comparator<Option<u32>>>,
}

impl Filter for UserFilter {
    fn new() -> Self {
        Self::default()
    }

    fn into_document(self) -> Result<Document, Error> {
        let mut doc = Document::new();
        if let Some(value) = self._id {
            doc.insert("_id", value);
        }
        if let Some(value) = self.name {
            doc.insert("name", ext::bson::Bson::try_from(value)?.0);
        }
        if let Some(value) = self.age {
            doc.insert("age", ext::bson::Bson::try_from(value)?.0);
        }
        Ok(doc)
    }
}

impl AsFilter<UserFilter> for User {
    fn filter() -> UserFilter {
        UserFilter::default()
    }

    fn into_filter(self) -> UserFilter {
        UserFilter {
            _id: None,
            name: Some(Comparator::Eq(self.name)),
            age: Some(Comparator::Eq(self.age)),
        }
    }
}

#[tokio::main]
async fn main() {
    // Create and async client
    let client = mongod::Client::new();

    // Insert a user into the users collection
    let user = User {
        name: "foo".to_string(),
        age: None,
    };
    let oid = client.insert_one::<User>(user).await.unwrap();
    println!("{:?}", oid);

    // Fetch all users in the users collection
    let mut cursor = client.find::<User, _>(None).await.unwrap();
    while let Some(res) = cursor.next().await {
        if let Ok((id, user)) = res {
            println!("{} - {:?}", id, user);
        }
    }

    // Update the user
    let mut filter = User::filter();
    filter.name = Some(Comparator::Eq("foo".to_owned()));
    let mut update = User::update();
    update.age = Some(Some(0));
    let updates = Updates {
        set: Some(update),
        ..Default::default()
    };
    let updated = client.update::<User, _, _>(filter, updates).await.unwrap();
    println!("updated {} documents", updated);

    // Find the user by oid
    let mut filter = User::filter();
    filter._id = Some(oid);
    let mut cursor = client.find::<User, _>(None).await.unwrap();
    while let Some(res) = cursor.next().await {
        if let Ok((id, user)) = res {
            println!("{} - {:?}", id, user);
        }
    }

    // Delete all users
    let deleted = client.delete::<User, _>(None).await.unwrap();
    println!("delete {} documents", deleted);
}
```

**Comment by mzdk100** [comment]
This is helpful to me, thank you.

**Comment by mzdk100** [comment]
Can #[mongo(filter)] be used to automatically implement AsFilter trait in future versions?

**Comment by alexkornitzer** [comment]
Yep, i'll try and get something out this weekend, i'll probably do something like `#[mongo(filter, oid)]`.

**Comment by alexkornitzer** [comment]
Implemented as `#[mongo(filter, oid)]` in `v0.3.5`. Also fixed a bug preventing `_id: Option<ObjectId>` from being added to a struct due to BSON derive.

**Comment by mzdk100** [comment]
I have used it perfectly!
